### ISSUE TITLE
files missing after sync

### ISSUE THREAD

**Issue by mapfeld** [Progeny42/refrapt#29]
**Describe the bug:**
Some files are missing after sync. I cannot find anything in the logs regarded to this issue. Restarting refrapt does not cure this issue.
```
root@computer:# LC_ALL=C apt dist-upgrade 
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Calculating upgrade... Done
The following NEW packages will be installed:
  linux-headers-5.10.0-16-amd64 linux-headers-5.10.0-16-common linux-image-5.10.0-16-amd64
The following packages have been kept back:
  odbcinst
The following packages will be upgraded:
  bash code dbeaver-ce dirmngr distro-info-data dpkg dpkg-dev firefox firefox-l10n-de gir1.2-nm-1.0 gnupg gnupg-l10n gnupg-utils gpg gpg-agent gpg-wks-client gpg-wks-server
  gpgconf gpgsm gpgv hexchat hexchat-common intel-microcode libdpkg-perl libfreetype6 libfribidi0 libgnutls30 libnm0 libsdl2-2.0-0 libsmbclient libwbclient0
  linux-compiler-gcc-10-x86 linux-headers-amd64 linux-image-amd64 linux-kbuild-5.10 linux-libc-dev logrotate msodbcsql17 mssql-tools nano network-manager
  network-manager-config-connectivity-debian openssh-client openssh-server openssh-sftp-server orca python3-samba samba-common samba-common-bin samba-libs signal-desktop
  smbclient tcpdump thunderbird thunderbird-l10n-de tigervnc-common tigervnc-standalone-server unrar usb.ids wireless-regdb
60 upgraded, 3 newly installed, 0 to remove and 1 not upgraded.
Need to get 102 MB/540 MB of archives.
After this operation, 384 MB of additional disk space will be used.
Do you want to continue? [Y/n] 
Err:1 http://computer/debian bullseye/main amd64 bash amd64 5.1-2+deb11u1
  404  Not Found [IP: 10.0.0.1 80]
Err:2 http://computer/debian bullseye/main amd64 dpkg amd64 1.20.11
  404  Not Found [IP: 10.0.0.1 80]
Err:3 http://computer/debian bullseye/main amd64 libgnutls30 amd64 3.7.1-5+deb11u1
  404  Not Found [IP: 10.0.0.1 80]
Err:4 http://computer/debian bullseye/main amd64 smbclient amd64 2:4.13.13+dfsg-1~deb11u4
  404  Not Found [IP: 10.0.0.1 80]
Err:5 http://computer/debian bullseye/main amd64 samba-common-bin amd64 2:4.13.13+dfsg-1~deb11u4
  404  Not Found [IP: 10.0.0.1 80]
Err:6 http://computer/debian bullseye/main amd64 python3-samba amd64 2:4.13.13+dfsg-1~deb11u4
  404  Not Found [IP: 10.0.0.1 80]
Err:7 http://computer/debian bullseye/main amd64 libsmbclient amd64 2:4.13.13+dfsg-1~deb11u4
  404  Not Found [IP: 10.0.0.1 80]
Err:8 http://computer/debian bullseye/main amd64 samba-libs amd64 2:4.13.13+dfsg-1~deb11u4
  404  Not Found [IP: 10.0.0.1 80]
Err:9 http://computer/debian bullseye/main amd64 libwbclient0 amd64 2:4.13.13+dfsg-1~deb11u4
  404  Not Found [IP: 10.0.0.1 80]
Ign:10 http://computer/debian bullseye/main amd64 samba-common all 2:4.13.13+dfsg-1~deb11u4
Ign:11 http://computer/debian bullseye/main amd64 distro-info-data all 0.51+deb11u2
Ign:12 http://computer/debian bullseye/main amd64 dpkg-dev all 1.20.11
Ign:13 http://computer/debian bullseye/main amd64 libdpkg-perl all 1.20.11
Err:14 http://computer/debian bullseye/main amd64 libfreetype6 amd64 2.10.4+dfsg-1+deb11u1
  404  Not Found [IP: 10.0.0.1 80]
Err:15 http://computer/debian bullseye/main amd64 libfribidi0 amd64 1.0.8-2+deb11u1
  404  Not Found [IP: 10.0.0.1 80]
Err:16 http://computer/debian bullseye/main amd64 logrotate amd64 3.18.0-2+deb11u1
  404  Not Found [IP: 10.0.0.1 80]
Err:17 http://computer/debian bullseye/main amd64 nano amd64 5.4-2+deb11u1
  404  Not Found [IP: 10.0.0.1 80]
Err:18 http://computer/debian bullseye/main amd64 openssh-sftp-server amd64 1:8.4p1-5+deb11u1
  404  Not Found [IP: 10.0.0.1 80]
Err:19 http://computer/debian bullseye/main amd64 openssh-server amd64 1:8.4p1-5+deb11u1
  404  Not Found [IP: 10.0.0.1 80]
Err:20 http://computer/debian bullseye/main amd64 openssh-client amd64 1:8.4p1-5+deb11u1
  404  Not Found [IP: 10.0.0.1 80]
Err:21 http://computer/debian bullseye/main amd64 network-manager amd64 1.30.6-1+deb11u1
  404  Not Found [IP: 10.0.0.1 80]
Err:22 http://computer/debian bullseye/main amd64 libnm0 amd64 1.30.6-1+deb11u1
  404  Not Found [IP: 10.0.0.1 80]
Err:23 http://computer/debian bullseye/main amd64 gir1.2-nm-1.0 amd64 1.30.6-1+deb11u1
  404  Not Found [IP: 10.0.0.1 80]
Err:24 http://computer/debian bullseye/main amd64 hexchat amd64 2.14.3-6+deb11u1
  404  Not Found [IP: 10.0.0.1 80]
Ign:25 http://computer/debian bullseye/main amd64 hexchat-common all 2.14.3-6+deb11u1
Err:26 http://computer/debian-security bullseye-security/non-free amd64 intel-microcode amd64 3.20220510.1~deb11u1
  404  Not Found [IP: 10.0.0.1 80]
Err:27 http://computer/debian bullseye/main amd64 libsdl2-2.0-0 amd64 2.0.14+dfsg2-3+deb11u1
  404  Not Found [IP: 10.0.0.1 80]
Err:28 http://computer/debian bullseye/main amd64 linux-compiler-gcc-10-x86 amd64 5.10.127-1
  404  Not Found [IP: 10.0.0.1 80]
Ign:29 http://computer/debian bullseye/main amd64 linux-headers-5.10.0-16-common all 5.10.127-1
Err:30 http://computer/debian bullseye/main amd64 linux-kbuild-5.10 amd64 5.10.127-1
  404  Not Found [IP: 10.0.0.1 80]
Err:31 http://computer/debian bullseye/main amd64 linux-headers-5.10.0-16-amd64 amd64 5.10.127-1
  404  Not Found [IP: 10.0.0.1 80]
Err:32 http://computer/debian bullseye/main amd64 linux-headers-amd64 amd64 5.10.127-1
  404  Not Found [IP: 10.0.0.1 80]
Err:33 http://computer/debian bullseye/main amd64 linux-image-5.10.0-16-amd64 amd64 5.10.127-1
  404  Not Found [IP: 10.0.0.1 80]
Err:34 http://computer/debian bullseye/main amd64 linux-image-amd64 amd64 5.10.127-1
  404  Not Found [IP: 10.0.0.1 80]
Err:35 http://computer/debian bullseye/main amd64 linux-libc-dev amd64 5.10.127-1
  404  Not Found [IP: 10.0.0.1 80]
Ign:36 http://computer/debian bullseye/main amd64 network-manager-config-connectivity-debian all 1.30.6-1+deb11u1
Ign:37 http://computer/debian bullseye/main amd64 orca all 3.38.2-2
Err:38 http://computer/debian bullseye/main amd64 tcpdump amd64 4.99.0-2+deb11u1
  404  Not Found [IP: 10.0.0.1 80]
Err:39 http://computer/debian bullseye/main amd64 tigervnc-standalone-server amd64 1.11.0+dfsg-2+deb11u1
  404  Not Found [IP: 10.0.0.1 80]
Err:40 http://computer/debian bullseye/main amd64 tigervnc-common amd64 1.11.0+dfsg-2+deb11u1
  404  Not Found [IP: 10.0.0.1 80]
Err:41 http://computer/debian bullseye/non-free amd64 unrar amd64 1:6.0.3-1+deb11u1
  404  Not Found [IP: 10.0.0.1 80]
Ign:42 http://computer/debian bullseye/main amd64 usb.ids all 2022.05.20-0+deb11u1
Ign:43 http://computer/debian bullseye/main amd64 wireless-regdb all 2022.04.08-2~deb11u1
Err:10 http://computer/debian bullseye/main amd64 samba-common all 2:4.13.13+dfsg-1~deb11u4
  404  Not Found [IP: 10.0.0.1 80]
Err:11 http://computer/debian bullseye/main amd64 distro-info-data all 0.51+deb11u2
  404  Not Found [IP: 10.0.0.1 80]
Err:12 http://computer/debian bullseye/main amd64 dpkg-dev all 1.20.11
  404  Not Found [IP: 10.0.0.1 80]
Err:13 http://computer/debian bullseye/main amd64 libdpkg-perl all 1.20.11
  404  Not Found [IP: 10.0.0.1 80]
Err:25 http://computer/debian bullseye/main amd64 hexchat-common all 2.14.3-6+deb11u1
  404  Not Found [IP: 10.0.0.1 80]
Err:29 http://computer/debian bullseye/main amd64 linux-headers-5.10.0-16-common all 5.10.127-1
  404  Not Found [IP: 10.0.0.1 80]
Err:36 http://computer/debian bullseye/main amd64 network-manager-config-connectivity-debian all 1.30.6-1+deb11u1
  404  Not Found [IP: 10.0.0.1 80]
Err:37 http://computer/debian bullseye/main amd64 orca all 3.38.2-2
  404  Not Found [IP: 10.0.0.1 80]
Err:42 http://computer/debian bullseye/main amd64 usb.ids all 2022.05.20-0+deb11u1
  404  Not Found [IP: 10.0.0.1 80]
Err:43 http://computer/debian bullseye/main amd64 wireless-regdb all 2022.04.08-2~deb11u1
  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/b/bash/bash_5.1-2%2bdeb11u1_amd64.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/d/dpkg/dpkg_1.20.11_amd64.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/g/gnutls28/libgnutls30_3.7.1-5%2bdeb11u1_amd64.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/s/samba/smbclient_4.13.13%2bdfsg-1%7edeb11u4_amd64.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/s/samba/samba-common-bin_4.13.13%2bdfsg-1%7edeb11u4_amd64.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/s/samba/python3-samba_4.13.13%2bdfsg-1%7edeb11u4_amd64.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/s/samba/libsmbclient_4.13.13%2bdfsg-1%7edeb11u4_amd64.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/s/samba/samba-libs_4.13.13%2bdfsg-1%7edeb11u4_amd64.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/s/samba/libwbclient0_4.13.13%2bdfsg-1%7edeb11u4_amd64.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/s/samba/samba-common_4.13.13%2bdfsg-1%7edeb11u4_all.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/d/distro-info-data/distro-info-data_0.51%2bdeb11u2_all.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/d/dpkg/dpkg-dev_1.20.11_all.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/d/dpkg/libdpkg-perl_1.20.11_all.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/f/freetype/libfreetype6_2.10.4%2bdfsg-1%2bdeb11u1_amd64.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/f/fribidi/libfribidi0_1.0.8-2%2bdeb11u1_amd64.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/l/logrotate/logrotate_3.18.0-2%2bdeb11u1_amd64.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/n/nano/nano_5.4-2%2bdeb11u1_amd64.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/o/openssh/openssh-sftp-server_8.4p1-5%2bdeb11u1_amd64.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/o/openssh/openssh-server_8.4p1-5%2bdeb11u1_amd64.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/o/openssh/openssh-client_8.4p1-5%2bdeb11u1_amd64.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/n/network-manager/network-manager_1.30.6-1%2bdeb11u1_amd64.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/n/network-manager/libnm0_1.30.6-1%2bdeb11u1_amd64.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/n/network-manager/gir1.2-nm-1.0_1.30.6-1%2bdeb11u1_amd64.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/h/hexchat/hexchat_2.14.3-6%2bdeb11u1_amd64.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/h/hexchat/hexchat-common_2.14.3-6%2bdeb11u1_all.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian-security/pool/updates/non-free/i/intel-microcode/intel-microcode_3.20220510.1%7edeb11u1_amd64.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/libs/libsdl2/libsdl2-2.0-0_2.0.14%2bdfsg2-3%2bdeb11u1_amd64.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/l/linux/linux-compiler-gcc-10-x86_5.10.127-1_amd64.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/l/linux/linux-headers-5.10.0-16-common_5.10.127-1_all.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/l/linux/linux-kbuild-5.10_5.10.127-1_amd64.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/l/linux/linux-headers-5.10.0-16-amd64_5.10.127-1_amd64.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/l/linux-signed-amd64/linux-headers-amd64_5.10.127-1_amd64.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/l/linux-signed-amd64/linux-image-5.10.0-16-amd64_5.10.127-1_amd64.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/l/linux-signed-amd64/linux-image-amd64_5.10.127-1_amd64.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/l/linux/linux-libc-dev_5.10.127-1_amd64.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/n/network-manager/network-manager-config-connectivity-debian_1.30.6-1%2bdeb11u1_all.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/o/orca/orca_3.38.2-2_all.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/t/tcpdump/tcpdump_4.99.0-2%2bdeb11u1_amd64.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/t/tigervnc/tigervnc-standalone-server_1.11.0%2bdfsg-2%2bdeb11u1_amd64.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/t/tigervnc/tigervnc-common_1.11.0%2bdfsg-2%2bdeb11u1_amd64.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/non-free/u/unrar-nonfree/unrar_6.0.3-1%2bdeb11u1_amd64.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/u/usb.ids/usb.ids_2022.05.20-0%2bdeb11u1_all.deb  404  Not Found [IP: 10.0.0.1 80]
E: Failed to fetch http://computer/debian/pool/main/w/wireless-regdb/wireless-regdb_2022.04.08-2%7edeb11u1_all.deb  404  Not Found [IP: 10.0.0.1 80]
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
```

**Configuration File and Logs:**
Provide the configuration file that you are using, and any logs available in the ~/refrapt/var directory. These will help to diagnose the issue.
Can provide on request. Or please tell me what to look for.

**System Information:**
OS: Debian Bullseye
Refrapt version: refrapt, version 0.4.4

**Comment by Progeny42** [comment]
Please provide your configuration file so I can attempt to replicate the issue

**Comment by mapfeld** [comment]
Please find attached the conf file.

[refrapt.txt](https://github.com/Progeny42/refrapt/files/9102509/refrapt.txt)

**Comment by Progeny42** [comment]
@rogerfachini It looks like you are right about your issue. Your log shows that incomplete downloads are being removed, as well as the cleaning process not finishing before the next schedule starts. As the initial download is so large, your scheduled jobs are running in parallel, interefering with the process. I can implement a feature so as to ensure only one instance of the program can run at any one time - this will allow you to use cron jobs from the get go, instead of manually downloading all items, and then setting up a cron job.

With regards @mapfeld's issue, unless they have also setup a cron job, it sounds like something else. From what they've said, they are letting refrapt run to completion. @mapfeld Have you performed the initial mirror manually, or only scheduled a job with cron?

**Comment by mapfeld** [comment]
@Progeny42 the program runs as cron job, but simultaneous execution is prevented by a lock file mechanism.

**Comment by Progeny42** [comment]
Ah I see. The Lock file that Refrapt generates is not actually to prevent simultaneous execution, but instead it is used as a signal for the next time it runs to indicate that the previous run was interrupted, either due to a crash, or from the job being cancelled (for example if you were running it via the command line). Due to the multithreaded nature of the program, catching and handling any crashes is very difficult, so it is not easily done to remove the lock file under abnormal circumstances.

**Comment by Progeny42** [comment]
I've just pushed a new release v0.4.5, which now makes Refrapt a singleton. This will only allow one instance to run at a time, and hopefully should resolve your issue.

To update perform the following:
```
python3.9 -m pip uninstall refrapt
python3.9 -m pip install refrapt
refrapt --version
```
The output should show that you have 0.4.5 installed.

If this fixes your problem, please close this issue.

**Comment by mapfeld** [comment]
Thank you for your efforts. I will check that. It may take some time, as the error usually occurred only after several days.

**Comment by mapfeld** [comment]
After letting the tool being run as cron job for a while now, today the following issue appeared when running `apt update:`
```
Hit:1 http://computer/dbeaver/debs/dbeaver-ce  InRelease
Hit:2 http://computer/microsoft/ubuntu/18.04/mssql-server-2019 bionic InRelease
Hit:3 http://computer/microsoft/debian/10/prod buster InRelease
Ign:4 http://computer/mint/packages elsie InRelease
Get:5 http://computer/debian bullseye InRelease [116 kB]
Hit:6 http://computer/debian bullseye-updates InRelease
Hit:7 http://computer/debian-security bullseye-security InRelease
Hit:8 http://computer/signal xenial InRelease
Hit:9 http://computer/microsoft/repos/code stable InRelease
Hit:10 http://computer/mint/packages elsie Release
Get:11 http://computer/debian bullseye-backports InRelease [49.0 kB]
Get:13 http://computer/debian bullseye-backports/main amd64 Packages.diff/Index [63.3 kB]
Get:14 http://computer/debian bullseye-backports/main i386 Packages.diff/Index [63.3 kB]
Get:15 http://computer/debian bullseye-backports/main Translation-en.diff/Index [63.3 kB]
Get:16 http://computer/debian bullseye-backports/main all Contents (deb).diff/Index [63.6 kB]
Ign:17 http://computer/debian bullseye-backports/main i386 Contents (deb).diff/Index
Ign:18 http://computer/debian bullseye-backports/main amd64 Packages T-2022-07-22-0203.10-F-2022-07-21-0817.16.pdiff
Ign:19 http://computer/debian bullseye-backports/main i386 Packages T-2022-07-21-2009.57-F-2022-07-19-2008.23.pdiff
Ign:20 http://computer/debian bullseye-backports/main Translation-en T-2022-07-21-2009.57-F-2022-07-21-2009.57.pdiff
Ign:21 http://computer/debian bullseye-backports/main all Contents (deb) T-2022-07-21-2009.57-F-2022-07-21-0817.16.pdiff
Get:17 http://computer/debian bullseye-backports/main i386 Contents (deb).diff/Index [63.6 kB]
Ign:17 http://computer/debian bullseye-backports/main i386 Contents (deb).diff/Index
Ign:22 http://computer/debian bullseye-backports/main amd64 Packages
Ign:23 http://computer/debian bullseye-backports/main i386 Packages
Ign:24 http://computer/debian bullseye-backports/main Translation-en
Ign:25 http://computer/debian bullseye-backports/main all Contents (deb)
Ign:26 http://computer/debian bullseye-backports/main i386 Contents (deb)
Get:22 http://computer/debian bullseye-backports/main amd64 Packages [320 kB]
Get:23 http://computer/debian bullseye-backports/main i386 Packages [319 kB]
Get:24 http://computer/debian bullseye-backports/main Translation-en [247 kB]
Get:25 http://computer/debian bullseye-backports/main all Contents (deb) [3876 kB]
Get:26 http://computer/debian bullseye-backports/main i386 Contents (deb) [1294 kB]
Err:26 http://computer/debian bullseye-backports/main i386 Contents (deb)
  File has unexpected size (1288161 != 1293629). Mirror sync in progress? [IP: 10.0.206.10 80]
  Hashes of expected file:
   - Filesize:1293629 [weak]
   - SHA256:eb83259bc2863d42d18143e6235dd36a9eadd8b0ad4ba2827bd23b6bf6d1c0be
  Release file created at: Sun, 24 Jul 2022 14:20:30 +0000
Fetched 4926 kB in 2s (3197 kB/s)
Reading package lists... Done
E: Failed to fetch http://computer/debian/dists/bullseye-backports/main/Contents-i386.gz  File has unexpected size (1288161 != 1293629). Mirror sync in progress? [IP: 10.0.206.10 80]
   Hashes of expected file:
    - Filesize:1293629 [weak]
    - SHA256:eb83259bc2863d42d18143e6235dd36a9eadd8b0ad4ba2827bd23b6bf6d1c0be
   Release file created at: Sun, 24 Jul 2022 14:20:30 +0000
E: Some index files failed to download. They have been ignored, or old ones used instead.
```

**Comment by Progeny42** [comment]
Assuming that your download has now completed fully at least once, this should not be occurring. This is saying that the Index file downloaded by refrapt does not match the expected size, based on another index file. All your Index files should be up to date. Best advice I can give is to ensure that;

1. Your log indicates that a Refrapt run has completed (look for either `No files eligible to clean` or `X bytes in Y files and directories will be freed...`)
2. Delete the `Contents-i386.gz` file, and do another download using Refrapt.

If you could send me your latest `refrapt.log`, I can take a look to see if there is anything obvious. Refrapt should be taking care to ensure that you always have the latest file, as it always tries to download the file if they are different sizes, like in your case, but it's possible there were some partial downloads as a result of overlapping cron jobs interfering with each other (though again, this should be accounted for).

**Comment by mapfeld** [comment]
I found the line you mentioned after a while searching for the relevant log file (expected it to be folder defined in varPath, but was aside of refrapt.conf).
Looking at this file, it seems as if there are runs that take several days (!!) to complete, the last one is running since `2022-07-25 21:00`. Shouldn't the copy mechanism with the `skel` folder prevent the problems from happening, in case this is the cause for the issues?

(Note: I provided the log file, but I flushed the folders somewhen in between.)
[refrapt.log](https://github.com/Progeny42/refrapt/files/9206191/refrapt.log)

**Comment by Progeny42** [comment]
Ok, I've had a look at your log file, and it still looks like there is overlapping of processes going on. For example, near the end of the log, you have
`[2022-07-25 11:23:14] INFO: Downloading 15063 Archive files...`
and then
`[2022-07-25 21:00:01] INFO: Starting Refrapt process`
without a
`INFO: Refrapt completed in X:XX:XX` happening inbetween.

Please try disabling your scheduled `cron` job, and then run refrapt manually. This will take time, as your log appears to show that most of your download has not actually occurred yet, due to the constant interruptions. Once the initial download is complete, then you can re-enable your `cron` job.

**Comment by mapfeld** [comment]
The problem [here](https://github.com/Progeny42/refrapt/issues/29#issuecomment-1193795038) persists. The logfile ends properly with the `[2022-08-12 01:09:25] INFO: Refrapt completed in 4:09:24` comment. The `apt update` was clearly executed after refrapt terminated.
Any further ideas about how to proceed?

**Comment by Progeny42** [comment]
Hmm okay. My initial suggestion would be:

1. Find and delete this file `computer/debian/dists/bullseye-backports/main/Contents-i386.gz`.
2. Use `set disableClean = True` in the configuration file
3. Run refrapt
4. Perform an `apt-get update` on your target machine

**Comment by mapfeld** [comment]
Thank you for your suggestions. Now, the `apt update` succeeds, but some of the packages are not found, e.g.
```
Err:1 http://computer/debian bullseye/main amd64 html2text amd64 1.3.2a-28
  404  Not Found [IP: 10.0.0.1 80]
Err:2 http://computer/debian bullseye/main amd64 gir1.2-graphene-1.0 amd64 1.10.4+dfsg1-1
  404  Not Found [IP: 10.0.0.1 80]
Err:3 http://computer/debian bullseye/main amd64 libxdo3 amd64 1:3.20160805.1-4
  404  Not Found [IP: 10.0.0.1 80]
Err:4 http://computer/debian bullseye/main amd64 libgsound0 amd64 1.0.2-5
  404  Not Found [IP: 10.0.0.1 80]
Err:5 http://computer/debian bullseye/main amd64 gir1.2-gsound-1.0 amd64 1.0.2-5
  404  Not Found [IP: 10.0.0.1 80]
```

**Comment by Progeny42** [comment]
I will use your configuration to download that repository, and see what happens. Will get back to you when I can.

**Comment by Progeny42** [comment]
Well, everything seemed to download ok. I have each of thise files from your snapshot on my filesystem from the download.

Just as a sanity check, if you manually go to the directory `/var/spool/apt-mirror/deb.debian.org/debian/pool/main/h/` there is no `html2text` directory, and no `html2text_1.3.2a-28_amd64.deb` file within it?

Could you please attach any `Archive-log.*` files from your `varPath` (`/var/spool/apt-mirror/var2`). I will check if there are any HTTP issues occurring on your end.

**Comment by mapfeld** [comment]
Hey @Progeny42, sorry for having such a long-lasting issue and thank you for your endurance.
The location you supplied has a text2html folder, but no file in it.
Solely the file [Archive-log.7.txt](https://github.com/Progeny42/refrapt/files/9351661/Archive-log.7.txt) has content, file Archive-log.8 is empty.

**Comment by Progeny42** [comment]
Equally so, I appreciate you being patient about this. Fully understand it can be frustrating when software does not work.

Those logs files were unfortunately not helpful in this instance. Could you go into your config file, and set the `logLevel` to `DEBUG`. Next time refrapt runs, could you send me the `refrapt.log`, as well as any `Archive-log.*` files again please? It should provide all the details about the process. It should say whether it attempts to download files like `html2text`, or whether it deems it unnecessary. Should help track this issue down.

**Comment by mapfeld** [comment]
You are welcome! Please find attached the [Archive-log.15](https://github.com/Progeny42/refrapt/files/9391747/Archive-log.15.txt) and [Archive-log.16](https://github.com/Progeny42/refrapt/files/9391748/Archive-log.16.txt), as well as the [refrapt.log](https://github.com/Progeny42/refrapt/files/9391749/refrapt.log) of the last two executions, one of those with `logLevel = DEBUG`. Grepping through all files in the `refrapt` and `varPath` folders does not yield any result…

**Comment by Progeny42** [comment]
Thanks for that. At the very least, it confirms it is not trying to download things like `html2text`, when it clearly should be. I will do another download of the repository, and then delete the .deb file, but keep the directory, and see what happens. There must be something up with one of the comparisons in the code.

**Comment by mapfeld** [comment]
OK, awesome! Please let me know once you have submitted the code. I'll test it as soon as I can.

**Comment by Progeny42** [comment]
Right, I see the issue now. The Packages.* file (Index file) is checked to see whether it has changed since the last time it was downloaded. As it has not, none of the files are eligible for download. As you initially started using this program with some `refrapt` runs being interrupted, you have parts of the repository downloaded, but have the latest Index file. Therefore, the files missing are not downloaded.

Your configuration has the `forceUpdate` setting to `True`, but this is not being checked in the correct place in the code, so it's not actually forcing the update. The `forceUpdate` setting only works to prevent files with the same size not being considered for download, whereas it needs to be applied earlier in the process to override the Index file check.

To test, I deleted the `html2text` debian file. Currently, if I run `refrapt`, it looks at the Index file, and says, the local version matches the downloaded version, so no work to do. With the fix in place, and the `forceUpdate` setting to `True`, the Index file is processed anyway, and all possible files are added for download. Wget will prevent most of the downloads, because the files are unchanged (because the Index file has not changed), but it will now download any missing files. After it finished, `html2text` now exists again.

This should be pretty quick. For example, my initial download of the 1 repository of 84GB took 2:30 hours. The fixed forced update only took 2 minutes.

What this change means is that, if `forceUpdate` is `True`, `refrapt` will compile every possible file, and try to download it every run. This is a bit excessive, and a bit slower, so I would advise you that once you have done a run with the new version with `forceUpdate` to `True` to get the remaining files, that you set the setting back to `False`. `refrapt` will then do its normal processing of only getting files if the Index file has changed.

Just finishing up testing, and will create a new release shortly.

**Comment by Progeny42** [comment]
Version `v0.4.6` is now live. Update as before, and then do the following:

1. Ensure the `forceUpdate` setting is `True` in your configuration.
2. Run `refrapt` - It will compile a list of all files to download, but `wget` will skip files that are already up to date, so it will not take as long.
3. Once the process is completed, and you've verified you have all the files now, change the `forceUpdate` setting to `False` for reduced processing times in the future, or keep it as is if you desire.
4. If you still have it set in your configuration, you can re-enable the cleaning process now with `set disableClean = False`.

If this addresses your problem, please close this Issue.

**Comment by mapfeld** [comment]
Honestly, I feel stupid at this point, but we are (or I am) back at the problem of post #1. The issue reoccurred after deleting the erroneous files and rerunning refrapt. It works immediately after, until 3 maybe days later. At least, running with `forceUpdate` runs way faster after the first iteration and fixed the issue.
I fear solving all that issues arising on my machine will turn your software into a paranoid file-checking Monk tool.

**Comment by Progeny42** [comment]
Sorry to hear it's still not working. I was sure we had it in that last update. Could you please send any `refrapt.log` files you have. I'll take a look and see if there is any strange behaviour still.

**Comment by mapfeld** [comment]
Well well. After the last iteration with a `forceUpdate`, everything works fine. For now, I would assume you really got it and something or someone just interfered with the execution (after all, the server is hosted by a third party out of my control). I would now assume it works until it proves me otherwise. My suggestion: we assume the problem being solved and I open a new ticket in case something does not work. Do you agree?

**Comment by Progeny42** [comment]
I'm happy to close the issue if you are. If you continue to have problems, we can investigate under a new issue. I will close this now. Best of luck.
